商品出品機能実装

rui4210 opened the pull request:
ご確認宜しくお願い致します。

↓ログイン画面
https://i.gyazo.com/2211db317fab73ddaf0279469e01e685.jpg
↓item単体テストコード
https://i.gyazo.com/056bbadd11b8b03074b2093a56b50284.png
↓商品登録　金額の Javascript
https://i.gyazo.com/bacfaf228bb3ad5b66be866b66a57580.png

rui4210 committed:
itemsテーブルの作成

rui4210 committed:
カテゴリー選択の実装

rui4210 committed:
商品出品機能実装

rui4210 committed:
itemモデル単体テストコード

tenpadrummer commented on app/controllers/application_controller.rb:
```diff
@@ -1,5 +1,5 @@
 class ApplicationController < ActionController::Base
-  before_action :authenticate_user!
+  #before_action :authenticate_user!
```
コメントアウトされたままのコードは削除しておきましょう！

rui4210 replied:
削除しました。

tenpadrummer commented on app/controllers/items_controller.rb:
```diff
@@ -0,0 +1,32 @@
+  before_action :move_to_index, except: [:index,]
+  def index
+    @items = Item.all
+    @items = Item.order("created_at DESC")
```
こちらのプルリクエストは、「商品出品機能の実装」でお間違いないでしょうか？
indexメソッドのアクションについて、このプルリクエストでレビューしてしまうと、今後のプルリクエストで拝見したいファイルが差分として上がってこない可能性があります。
これらは次回以降のプルリクエストで拝見させていただきたいので、一旦こちらは削除しておくといいでしょう。

rui4210 replied:
「商品出品機能の実装」で間違いありません。
indexメソッドを削除しました。

tenpadrummer commented on app/controllers/items_controller.rb:
```diff
@@ -0,0 +1,32 @@
+
+  def move_to_index
+    unless user_signed_in?
+      redirect_to '/users/sign_in'
```
'/users/sign_in'はprefixを使用しましょう！

rui4210 replied:
prefixを使用した形に修正しました。
↓rails routesの内容です。
https://i.gyazo.com/5c073755fa49bcd9d09f89de965f33cf.png

tenpadrummer commented on app/javascript/packs/application.js:
```diff
@@ -4,9 +4,11 @@
 
 require("@rails/ujs").start()
-require("turbolinks").start()
+// require("turbolinks").start()
```
コメントアウトされたままのコードは削除しておきましょう！

rui4210 replied:
削除しました。

rui4210 replied:
削除しました。

tenpadrummer commented on app/models/item.rb:
```diff
@@ -0,0 +1,53 @@
+  #発送元の地域の選択が「--」の時は保存できないようにする
+  validates :shipping_region_id, numericality: { other_than: 1 }
+  #発送までの日数の選択が「--」の時は保存できないようにする
+  validates :shipping_day_id, numericality: { other_than: 1 }
```
```
#商品情報
  validates :image, presence: true
  #商品名が必須であること
  validates :name, presence: true
  #商品の説明が必須であること
  validates :description, presence: true
  #カテゴリーの情報が必須であること
  validates :category_id, presence: true
  #商品の状態についての情報が必須であること
  validates :product_status_id, presence: true
  #配送料の負担についての情報が必須であること
  validates :shipping_charges_id, presence: true
  #発送元の地域についての情報が必須であること
  validates :shipping_region_id, presence: true
  #発送までの日数についての情報が必須であること
  validates :shipping_day_id, presence: true
  #価格についての情報が必須であること
  validates :price, presence: true
  #価格の範囲が、¥300~¥9,999,999の間であること
  validates :price, inclusion: { in: 300..9999999 }
  #販売価格は半角数字のみ入力可能であること
  VALID_PRICE_REGEX = /\A[0-9]+\z/
  validates :price, format: { with: VALID_PRICE_REGEX }
  #カテゴリーの選択が「--」の時は保存できないようにする
  validates :category_id, numericality: { other_than: 1 }
  #商品の状態の選択が「--」の時は保存できないようにする
  validates :product_status_id, numericality: { other_than: 1 }
  #配送料の負担の選択が「--」の時は保存できないようにする
  validates :shipping_charges_id, numericality: { other_than: 1 }
  #発送元の地域の選択が「--」の時は保存できないようにする
  validates :shipping_region_id, numericality: { other_than: 1 }
  #発送までの日数の選択が「--」の時は保存できないようにする
  validates :shipping_day_id, numericality: { other_than: 1 }
```

バリデーションは「with_options」を使用することで、共通したオプションを付けることが可能となります。したがって、こちらのカリキュラムの内容を参考にまとめてみてはいかがでしょう？https://master.tech-camp.in/curriculums/5001

rui4210 replied:
presence: trueとnumericality: { other_than: 1 }の２項目のバリデーションをwith_optionでまとめました。

tenpadrummer commented on app/views/items/index.html.erb:
```diff
@@ -182,7 +182,7 @@
 <div class='purchase-btn'>
   <span class='purchase-btn-text'>出品する</span>
-  <a href="#">
+  <a href="/items/new">
```
こちらaタグではなく、link_toとprefixを使用しましょう！

rui4210 replied:
linK_toとprefixを用いての記述に修正しました。
↓rails routesの内容です。
https://i.gyazo.com/8f0ad9fb2518484afb94eb7a65bb9ce2.png

tenpadrummer reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

rui4210 committed:
コードレビュー修正（１回目）

- prefixの記述に変更
- バリデーションでwith_opitonを使用
- コメントアウトをしている記述を削除
- コントローラーのdef index削除

syuji-yamamoto commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-    <%# render 'shared/error_messages', model: f.object %>
+    <%= render 'shared/error_messages', model: f.object %>
     <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
不要なコメントアウトは削除しておきましょう！

rui4210 replied:
上下のコメントアウトを削除しました。

syuji-yamamoto reviewed (commented):
レビュー依頼ありがとうございます！
細かい点ですが、コメントしましたので、ご確認ください！

rui4210 committed:
コードレビュー修正（２回目）

- コメントアウトを削除

kokikurahashi commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,89 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price can't be blank")
+      end
+      it "価格についての情報が¥300~¥9,999,999の範囲外の価格では登録できない" do
+        @item.price = '299,10000000'
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price is not included in the list")
+      end
+      it "価格についての情報が半角数字以外の入力だと登録できない" do
+        @item.price = '３００'
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price is not included in the list")
+      end
```
こちら価格のテストコードに対するエラー文が同一のものになっていますが、それぞれ別々のテストなのでそれぞれに対して適切なエラー文がでるようにitemモデルにエラーメッセージを設定し、それぞれテストので別のエラー文がでるようにしましょう！！

rui4210 replied:
modelsフォルダ内のitem.rb内で、:messageを使用してみました。

kokikurahashi reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました！
コメントしましたので、ご確認をお願いいたします！

rui4210 committed:
コードレビュー修正（３回目）

i-date commented on app/models/item.rb:
```diff
@@ -41,10 +41,10 @@ class Item < ApplicationRecord
 
   #価格の範囲が、¥300~¥9,999,999の間であること
-  validates :price, inclusion: { in: 300..9999999 }
+  validates :price, inclusion: { in: 300..9999999 }, presence: { message: 'price is not ¥300~¥9,999,999' }
```
44行目ですが、メッセージがinclusionではなくてpresensceに対して設定されていないでしょうか？

rui4210 replied:
inclusion内にmessage:を入れる修正をしました。

i-date commented on app/models/item.rb:
```diff
@@ -41,10 +41,10 @@ class Item < ApplicationRecord
   #販売価格は半角数字のみ入力可能であること
   VALID_PRICE_REGEX = /\A[0-9]+\z/
-  validates :price, format: { with: VALID_PRICE_REGEX }
+  validates :price, format: { with: VALID_PRICE_REGEX }, presence: { message: 'price is not half width digit' }
```
47行目も、メッセージがformatではなくてpresensceに対して設定されていないでしょうか？

rui4210 replied:
formatをnumericalityに変更しました。（価格の範囲のバリデーションを拾うエラーが出たので、メンターの方に相談の上変えました。）
numericality内にmessage:を入れる修正をしました。

i-date commented on spec/models/item_spec.rb:
```diff
@@ -77,12 +77,12 @@
         @item.price = '299,10000000'
         @item.valid?
-        expect(@item.errors.full_messages).to include("Price is not included in the list")
+        expect(@item.errors.full_messages)
```
こちらの80行目は、設定していただいたエラーメッセージを含んでいるか確認しましょう。

rui4210 replied:
.to include("Price is not ¥300~¥9,999,999")を追記し、修正しました。

i-date commented on spec/models/item_spec.rb:
```diff
@@ -77,12 +77,12 @@
+        @item.price = '３三さんサン'
         @item.valid?
-        expect(@item.errors.full_messages).to include("Price is not included in the list")
+        expect(@item.errors.full_messages)
```
こちらの85行目も、設定していただいたエラーメッセージを含んでいるか確認しましょう。

rui4210 replied:
.to include("Price is not half width digit")を追記し、修正しました。

i-date reviewed (commented):
修正ありがとうございます！
追加コメントいたしましたので、ご確認お願いいたします。

rui4210 committed:
コードレビュー修正（４回目）

- :message修正

i-date reviewed (approved):
修正ありがとうございます！
お疲れさまでした、LGTMとさせていただきます。
以降の実装も引き続きこの調子で進めていきましょう。応援しております！
